download/npmExternalToInternalRedirect/npmExternalToInternalRedirect.groovy

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright (C) 2014 JFrog Ltd.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import org.artifactory.repo.RepoPath
+import org.artifactory.repo.RepoPathFactory
+
+
+download {
+
+    altRemoteContent { repoPath ->
```

Comment:
I think this is the wrong hook. `altRemoteContent` triggers at the boundary between Artifactory and a remote location, not at the one between a user and Artifactory. One effect of this is that the alternate content is saved in the remote cache in Artifactory, which I can see causing problems. Say someone downloads a remote package `foo 3.4`, and then someone else uploads a local package `foo 1.0`. The next time someone downloads `foo`, they'll get `foo 3.4` from the remote repository cache, and the plugin won't run. Or say someone downloads `foo 1.0`, which gets cached, and then the local `foo 1.0` is removed. The next time someone downloads `foo`, they'll get `foo 1.0` from the cache instead of `foo 3.4`.

The `altResponse` hook seems more correct here, to me.

Reply:
The requirement for us was to always take the `foo` package in the internal repo. No external requests for foo would be permitted. In this case, if someone tries to specifically get `foo 3.4` they should get a 404 . In the above scenario the package `foo 3.4` will never be in the external cache. So in the external cache we will always have the latest version of local package.
Also i do not think that we will not be removing the packages from the local repository.

Another reason for choosing this hook was to minimise the number of calls to plugin and increase the response time for all other calls.

Reply:
>The requirement for us was to always take the `foo` package in the internal repo.

That's of course assuming a `foo` package exists in the internal repo; my first example was about if the remote `foo` package was cached before the first internal version of `foo` was added to the local repo. Beyond that, I agree with your logic, so it's fine if this stays as-is.

>if someone tries to specifically get `foo 3.4` they should get a 404 .

Good to know. In that case, some of the rest of the code is wrong; it seems to check specifically for the requested version, not just the package name. I'll take another look and see if I can make more specific comments.